front/components/members/InvitationModal.tsx

```diff
@@ -0,0 +1,524 @@
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+    },
+    body: JSON.stringify(body),
+  });
+
+  if (!res.ok) {
+    let data: any = {};
+    try {
+      data = await res.json();
+    } catch (e) {
+      // ignore
+    }
+    if (data?.error?.type === "invitation_already_sent_recently") {
+      sendNotification({
+        type: "error",
+        title: emails.length === 1 ? "Invite failed" : "Invites failed",
+        description:
+          (emails.length === 1 ? "This user has" : "These users have") +
+          " already been invited in the last 24 hours. Please wait before sending another invite.",
+      });
+    }
+
+    sendNotification({
+      type: "error",
+      title: "Invite failed",
+      description:
+        "Failed to invite new members to workspace: " + res.statusText,
+    });
+  } else {
+    const result: PostInvitationResponseBody = await res.json();
+    const failures = result.filter((r) => !r.success);
+
+    if (failures.length > 0) {
+      sendNotification({
+        type: "error",
+        title: "Some invites failed",
+        description:
+          result[0].error_message ||
+          `Failed to invite ${failures} new member(s) to workspace.`,
+      });
+    } else {
+      sendNotification({
+        type: "success",
+        title: "Invites sent",
+        description: isNewInvitation
+          ? `${emails.length} new invites sent.`
+          : `Sent ${emails.length} invites again.`,
+      });
+    }
+  }
+}
+
+async function updateInvitation({
+  owner,
+  invitation,
+  newRole,
+  mutate,
+  sendNotification,
+  confirm,
+}: {
+  owner: WorkspaceType;
+  invitation: MembershipInvitationType;
+  newRole?: RoleType; // Optional parameter for role change
+  mutate: any;
+  sendNotification: (notificationData: NotificationType) => void;
+  confirm?: (confirmData: ConfirmDataType) => Promise<boolean>;
+}) {
+  if (newRole) {
+    const r = await fetch(`/api/w/${owner.sId}/invitations/${invitation.sId}`, {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/json",
+      },
+      body: JSON.stringify({
+        status: invitation.status,
+        initialRole: newRole,
+      }),
+    });
+
+    if (!r.ok) {
+      const error: { error: { message: string } } = await r.json();
+      window.alert(error.error.message);
+    }
+    sendNotification({
+      title: "Success!",
+      description: "Invitation role successfully updated.",
+      type: "success",
+    });
+    await mutate(`/api/w/${owner.sId}/invitations`);
+  } else {
+    if (
+      !(
+        confirm &&
+        (await confirm({
+          title: "Revoke invitation",
+          message: `Are you sure you want to revoke the invitation for ${invitation.inviteEmail}?`,
+          validateLabel: "Yes, revoke",
+          validateVariant: "primaryWarning",
+        }))
+      )
+    ) {
+      return;
+    }
+
+    const res = await fetch(
```

Comment:
The two fetch can be factored as they are exactly the same no?

Reply:
My bad, needed one more ☕

Reply:
https://github.com/dust-tt/dust/pull/5275